### PR TITLE
Design document on result records

### PR DESCRIPTION
**Pull request by mih**
Developers are encouraged to amend this initial document with additional information, in particular field names that are in active use.

- Fixes datalad/datalad#6054
- Fixes datalad/datalad#6171

Please suggest concrete changes/addition using the reviewing functionality of by pushing the this branch. Given the documentation-nature of this PR, I will merge this tomorrow, unless objections are expressed.

### CODE REVIEW

File: docs/source/design/result_records.rst [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+- ``impossible``: a requested operation cannot be performed, possibly because
+  its preconditions are not met
+- ``error``: an error occurred while performing an operation
+
+Based on the ``status`` field, a result is categorized into *success* (``ok``,
+``notneeded``) and *failure* (``impossible``, ``error``). Depending on the
+``on_failure`` parameterization of a command call, any failure-result emitted
+by a command can lead to an ``IncompleteResultsError`` being raised on
+command exit, or a non-zero exit code on the command line.
+
```
Comment (adswa) [inline docs/source/design/result_records.rst:82]: In addition to raising errors, isn't it also "nothing" when we set ``on_failure='ignore'``? Could we maybe add the possible values that `on_failure` can take (sorry if I have missed them)?

**Comment by adswa** [comment]
I find this very helpful! Left a few comments

File: docs/source/design/result_records.rst (hunk 1 above)
Comment (mih) [reply docs/source/design/result_records.rst:82]: Will do, good point, thx!

File: docs/source/design/result_records.rst [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+------------------
+
+List of any error messages that were captured or produced while achieving a
+result.
+
+
+``exception``
+-------------
+
+An exception that occurred while achieving the reported result.
```
Comment (bpoldrack) [inline docs/source/design/result_records.rst:179]:
```suggestion
A ``CapturedException`` that occurred while achieving the reported result. If provided, ``get_status_dict`` will fill in the ``exception_traceback`` field automatically.
```

File: docs/source/design/result_records.rst [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+If a result record has a ``message`` field, then a given `Logger` instance
+(typically from ``logging.getLogger()``) will be used to automatically log
+this message. The log channel/level is determined based on
+``datalad.log.result-level`` configuration setting. By default, this is
+the ``debug`` level. When set to ``match-status`` the log level is determined
+based on the ``status`` field of a result record:
+
+- ``debug`` for ``'ok'``, and ``'notneeded'`` results
+- ``warning`` for ``'impossible`` results
+- ``error`` for ``'impossible`` results
```
[the remark this answers is not in the document]
Comment (mih) [reply docs/source/design/result_records.rst:133]: Thx, it also has a quote too much. Will fix.

File: docs/source/design/result_records.rst (hunk 2 above)
Comment (mih) [reply docs/source/design/result_records.rst:179]: Is that true? It seems the code takes any exception. The feature is undocumented.

Comment (bpoldrack) [reply docs/source/design/result_records.rst:186]: Result rendering may take any, but `get_status_dict` doesn't. So, at least when taking advantage of that convenience feature `CapturedException` would be required.

Comment (bpoldrack) [reply docs/source/design/result_records.rst:186]: FWIW: I'm fine with declaring that `get_status_dict` *should* be able to deal with any. However, that's defining current behavior as a bug and I'd argue it's still worth documenting that feature here.

Comment (mih) [reply docs/source/design/result_records.rst:186]: I have no desire to declare anything. I am documenting what I think I know is going on. It seems that this feature was exactly how I had it in mind and described it (added by @yarikoptic in https://github.com/datalad/datalad/commit/b737726673c4e47ad3e15d78654eb981146bbe79), but was silently changed (now restricted to `CapturedException`) by @bpoldrack with the recent https://github.com/datalad/datalad/commit/6d4f1ebb901010725b9dba4dd4d2e75af474dbfe.

**Pull request merged**